### PR TITLE
Tpetra fema updates

### PR DESCRIPTION
**Pull request by william76**
@trilinos/tpetra 
@mhoemmen 

## Description
Finite Mesh Assembly Total Element Loop (Type-2) fix. 

In the original code, we were always inserting owned+shared indices into the overlapping maps.  This fixes that issue and we insert owned into owned and shared into shared now.

I also cleaned up a little bit of output formatting in all the FEMA examples.

### CODE REVIEW

File: packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp [hunk 1]
```diff
@@ -284,8 +295,16 @@ int executeLocalElementLoopDP_(const comm_ptr_t& comm, const struct CmdLineOpts&
       }
 
       // For Type-2 Assembly, we only sumInot the overlapping crs_matrix.
```
Comment (mhoemmen) [inline packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp]: "sumInto" -- also I don't think the comment applies any more, so you could just delete it.

**Comment by william76** [comment]
@mhoemmen 
I'm seeing a segfault in the PR testing... in that test, the mesh size is 200x200 with 4 processes.  When I ran this locally using 100x100 it didn't error.  

If you get a chance, can you have a look and let me know if there was anything we missed in our edits?

**Comment by mhoemmen** [comment]
@william76 Try running with the following environment variable settings:
```
TPETRA_DEBUG=1
TPETRA_VERBOSE=1
```

**Comment by mhoemmen** [comment]
`Tpetra::MultiVector::sumIntoGlobalValue: Global row index 100is not present on this process 0.`

**Comment by mhoemmen** [comment]
@william76 You're asking the mesh if it owns the element node index.  Wouldn't you instead want to ask the (owned row) Map if it owns `global_row_id`?

**Comment by mhoemmen** [comment]
> BTW, do you know how to copy lines or ranges of code into github comments where it shows line numbers, etc?

View the file in the GitHub web interface, click on the line number, and "copy permalink."  For example:

https://github.com/trilinos/Trilinos/blob/a3875b6b40358a8915308b600258a85f0a952f73/packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp#L288

**Comment by mhoemmen** [comment]
https://github.com/trilinos/Trilinos/blob/a3875b6b40358a8915308b600258a85f0a952f73/packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp#L288-L290

**Comment by william76** [comment]
Yeah, just now caught that...  hmmm... for that matter, I wonder if I had the similar loop in the graph construction also wrong...

so, if I copy the permalink and past that I just get a link...
https://github.com/william76/Trilinos/blob/a3875b6b40358a8915308b600258a85f0a952f73/packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp#L183-L190

**Comment by william76** [comment]
@mhoemmen 
Looks like it's fixed, adding the `--save-mm` option and comparing the generated matrix with all three examples is working properly now.  I'll remove the WIP flag and let the autotester re-run the tests.

**Comment by mhoemmen** [comment]
> Perhaps the link must be only within the same repo... can't do it on a fork?

I think you have to "view" the file in the commit.

File: packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp [hunk 2]
```diff
@@ -174,15 +180,22 @@ int executeLocalElementLoopDP_(const comm_ptr_t& comm, const struct CmdLineOpts&
     //
     for(size_t element_node_idx=0; element_node_idx<owned_element_to_node_ids.extent(1); element_node_idx++)
     {
-       crs_graph_overlapping->insertGlobalIndices(global_ids_in_row[element_node_idx], global_ids_in_row());
+      if( mesh.nodeIsOwned(global_ids_in_row[element_node_idx]) )
```
Comment (mhoemmen) [inline packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp]: What does `element_node_idx` actually return?  Is that an element index?  If so, then this is all wrong -- you would need a second loop over rows of the element.

Comment (william76) [reply packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp]: If it's wrong would it be generating a bad mesh or would it do something more subtle?

Comment (william76) [reply packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_LocalElementLoop_DP.hpp]: @mhoemmen 
`element_node_idx` is an index into the `owned_element_to_node_ids` array.  It is not a global node index. Accessing `global_ids_in_row[element_node_idx]` gives me the global node id of my node.

for example, for a 1 process 3x3 test: "[pid] : (element_node_idx, global_ids_in_row[element_node_idx])":

```
element 0
[0] : (0 ,0)     mesh.nodeIsOwned( 0 ) => true
[0] : (1 ,1)     mesh.nodeIsOwned( 1 ) => true
[0] : (2 ,5)     mesh.nodeIsOwned( 5 ) => true
[0] : (3 ,4)     mesh.nodeIsOwned( 4 ) => true
element 1
[0] : (0 ,1)    .
[0] : (1 ,2)    .
[0] : (2 ,6)    .
[0] : (3 ,5)    .
element 2
[0] : (0 ,2)
[0] : (1 ,3)
[0] : (2 ,7)
[0] : (3 ,6)
element 3
[0] : (0 ,4)
[0] : (1 ,5)
[0] : (2 ,9)
[0] : (3 ,8)
element 4
[0] : (0 ,5)
[0] : (1 ,6)
[0] : (2 ,10)
[0] : (3 ,9)
element 5
[0] : (0 ,6)
[0] : (1 ,7)
[0] : (2 ,11)
[0] : (3 ,10)
element 6
[0] : (0 ,8)
[0] : (1 ,9)
[0] : (2 ,13)
[0] : (3 ,12)
element 7
[0] : (0 ,9)
[0] : (1 ,10)
[0] : (2 ,14)
[0] : (3 ,13)
element 8
[0] : (0 ,10)
[0] : (1 ,11)
[0] : (2 ,15)
[0] : (3 ,14)
```

**Comment by allevin** [comment]
@william76 FYI, the Automerge will not be performed unless you get an "approved" on your Pre-Merge inspection.  However that will still be a valid test on the Automerge feature.

**Comment by william76** [comment]
@allevin 
Yeah, I was hopeful to get the automerge tried out if @mhoemmen can approve ;)

**Comment by william76** [comment]
@allevin will automerge do the automerge if an approver approves changes after testing is completed or does it only work if the change is approved prior to test completions?

**Comment by allevin** [comment]
@william76 if the code is currently being tested and it is approve (and there are no new commits), the Automerge should work.  So if you can sucker @mhoemmen to approve while it's being tested... It should Automerge.

**Pull request merged**